### PR TITLE
Propose a name schema for counters, implement first four ingress counters

### PR DESCRIPTION
**Pull request by teknico**
Check the counters by changing to che `src/` directory and running:

```
$ make && sudo ./program/lwaftr/check/check_counters.sh
```

### CODE REVIEW

File: src/apps/lwaftr/lwaftr.lua [hunk 1]
```diff
@@ -205,8 +205,25 @@ function LwAftr:new(conf)
...
+   -- - size: "bytes", "packets".
+   -- Optional fields:
+   -- - handling: "drop", "reass"
+   -- - reason: reasons for dropping or reassembling (TBD);
+   -- Aggregated values (i.e. all sent, received, dropped packets) are computed
+   -- at request time by adding up all relevant counters.
+   local counters_dir = "app/lwaftr/counters/"
    o.counters = {}
-   o.counters.v4rxbytes = counter.open("lwaftr.v4/rxbytes")
+   o.counters.rx_ip_v4_bytes = counter.open(counters_dir.."rx-ip-v4-bytes")
```
Comment (kbara) [inline src/apps/lwaftr/lwaftr.lua]: I'd prefer spaces around the ''..', though we don't have a consistent style or style guide in this regard.

File: src/program/lwaftr/check/check_counters.sh [hunk 2]
```diff
@@ -0,0 +1,8 @@
+#! /usr/bin/env bash
+# Run with sudo from the src directory.
+./snabb lwaftr check \
```
Comment (kbara) [inline src/program/lwaftr/check/check_counters.sh]: By convention, I tend to put the input pcaps on one line and the outputs on another unless the line would get too long; it's not crucial, though.

**Comment by kbara** [comment]
LGTM iff the test passes.

**Comment by wingo** [comment]
I don't think we want separate tests to test the counters -- we want to annotate each individual test in end-to-end.sh and end-to-end-vlan.sh with their expected changes to counters.

**Comment by kbara** [comment]
Hopefully a subset of counters, or we'll have to update every test every time we add a relevant counter?

**Comment by wingo** [comment]
kbara: counters not present in the expectations file would be assumed not to change.  if we add a counter, it probably won't change most tests, if any.  but yeah for me i would like to test each end-to-end on all counters.

**Comment by teknico** [comment]
Making the requested changes. Any comments re: the filename schema?

**Comment by kbara** [comment]
Hmm. Probably something like test_description.counters makes more sense for the file with expected counter values; 'counters.lua' doesn't scale to most/all of the tests. (I think Andy's idea about counter coverage is probably right - it makes me twitch a little about test fragility, but I think he's making the right set of tradeoffs on it).

**Comment by wingo** [comment]
File name schema looked great to me!  Clean and orthogonal :)

**Comment by teknico** [comment]
@kbara, not sure what you mean by " 'counters.lua' doesn't scale to most/all of the tests". Would `test_description.counters` have a different format? And why would it scale better?

**Comment by kbara** [comment]
I'm reading counters.lua ( https://github.com/teknico/snabb/blob/four-counters-and-names/src/program/lwaftr/check/counters.lua ) as being per-test, since the current contents are 
```
return {
   rx_ip_v4_bytes = 66,
   rx_ip_v4_packets = 1,
   rx_ip_v6_bytes = 106,
   rx_ip_v6_packets = 1,
}
```
So, either I'm missing something (quite possible!) or you'll need a different but conceptually similar file per test, in which case they'll need distinct filenames.

**Comment by teknico** [comment]
Ah, I see. Yes, with the current structure a different counters file will be needed for each test, which is going to be unwieldy. One single counters data file would be better, with each test picking up its own counters data from it.

**Comment by teknico** [comment]
But I guess I'll worry about that in the next branch, if it's ok with you. :-)

**Comment by kbara** [comment]
Yep, fine by me, just mentioning since you asked for filename feedback. :)

**Comment by wingo** [comment]
I was thinking a separate counters expectations file for each test, like we have separate output pcap expectations for each tests; but open to anything :)

**Comment by teknico** [comment]
Well, maybe it's not that unwieldy, and implementation-wise it only requires changes in filenames and not in code, so let's see how it pans out.

**Comment by kbara** [comment]
Just thinking out loud here: WDYT about having 'check' set up to verify all the counters against an expected counter set which enumerates all non-zero counters, so it dies if any unspecified counters are non-zero? (Counters would be specified on a per-test basis)

Advantages:
- We won't miss necessary updates to new counters
- We won't miss if a counter we haven't specified is incremented if we thought it wouldn't be

Disadvantages:
- Sometimes this'll force a bunch of fragile updates - but that's probably better than hiding the implications of changes and hoping for the best

Acks, "maybe later", or replies of "that is a *terrible* idea for these reasons" all welcome. :-)

**Comment by kbara** [comment]
Re: changes in filenames and not in code - one general engineering principle I reflexively work on is the ideas that config / other peripheral non-code errors are *at least* as likely to cause catastrophic cascading failures as the actual code, so are similarly important to test/try to verify/etc, at least after initial prototyping phases. (While prototyping, a much more relaxed approach can be helpful!)

**Comment by teknico** [comment]
Currently the "master" is the `counters.lua` files (or whatever they're going to be called): only the expected values of the counters listed therein are checked against the actual values. IIUC, you propose to make the actual non-zero values to be the "master", and be all checked against the expected values listed in `counters.lua`.

It does seem a more comprehensive approach, but I wonder, why just the non-zero actual values? It would also be useful to make sure that the _zero_ actual values are also supposed to be zero.

So we could instead check all counters all the time, I guess, keeping the `counters.lua` files as "master", and listing all counters in all of them, even those (the large majority each time) that are going to have a zero actual value.

**Comment by kbara** [comment]
We're discussing almost the same thing; the question is whether the file explicitly contains the zero, or the system is designed so that everything unspecified MUST be zero or it's a failure. I'm leaning towards the latter approach, so that the files stay more readable and don't need to be changed every time a counter that's irrelevant to them is added.

We both agree that the zero values should be checked; it's a question of notation that we're noodling on differently. :-)

**Comment by teknico** [comment]
Indeed. In this case, "explicit is better than implicit" conflicts with "simple is better than complex": your way is slightly less explicit, but quite a bit simpler and less verbose. I'll give it a try. (Sorry for the Pythonisms. :-) )

**Comment by kbara** [comment]
*Pythonista handshake* ;-)

It's less explicit, but it reduces churn/noise if it's not too much of a PITA to implement - and we ideally need to have some mechanism for checking for unspecified counters either way.

**Comment by teknico** [comment]
All counters are defined both in `LwAftr:new` and in `check.lua`. The check code records changed values for all the defined counters: then those differences are checked against the `counters.lua` file. Any (unexpectedly) changed value for a counter not mentioned in `counters.lua` will be detected.

@wingo, once all counters are defined, I'll make one single pass through all existing standard and vlan tests. I'd like to avoid having to make several of those passes, if at all possible. :-) So I added a counters test to `end-to-end.sh` as a temporary measure.

**Comment by kbara** [comment]
The counters for the vlan tests *should* be identical (I think) or easily derived (if there were vlan-related counters); I'd treat them as identical and take any exceptions from that to be something to squint at and think about quite hard.

**Comment by teknico** [comment]
Yeah, diffing `end-to-end.sh` and `end-to-end-vlan.sh` is... interesting. :-)

**Comment by kbara** [comment]
If you want to change the variable naming scheme or do other things to make the diffs less noisy, feel free; I'm more than happy to review such changes if they save you any bother.

**Comment by kbara** [comment]
LGTM, with the caveat that I'll want a different name for counters.lua before it hits our upstream (Igalia repo) lwaftr branch. Thanks for your work on it!

**Pull request merged**